### ISSUE TITLE
reuse client when using tokens

### ISSUE THREAD

**Issue by mdzieg** [Azure/autorest#3192]
How to update credentials (TokenCredentials)?
I would like to use the same instance of the client (reuse HttpClient). I can do that unless I have to update token.

Credentials setter is private.

Any hints?
Maybe Credentials could be public?

**Comment by JamieTemple** [comment]
Hello,

I would like to second this, but suspect we are using different clients - in my case it's TypeScript.

I have an Angular app that is using my autorest generated client.

I've set up the app to DI the service into my components - everything is looking good at this stage.

... that is, until the access token gets refreshed, then the instance of the service that was DI'd into my components constructor now has an out-of-date access token :(

The MyServiceContext class generated by autorest includes a `credentials` property, which made me think "hey, this is easy, I just need to refresh the credentials when my token gets silently refreshed" ... so I stuck the following into my components constructor:

```
    constructor(private myService: MyService, private oauthService: OAuthService) {
        // Keep our auth tokens up to date.
        this.oauthService.events.subscribe(event => {
            if (event.type == "silently_refreshed") {
                this.myService.credentials = new TokenCredentials(this.oauthService.getAccessToken());
            }
        });
    }

```

Bang, job done! ... but no :(

... so it seems that even though the generated MyServiceContext class has a `credentials` property, it's not used beyond the constructor.

... after a little digging, it looks like the generated code inherits from an `msRest.ServiceClient`, where the credentials you use to instantiate your class are used in the `ServiceClient` constructor to build an array of `requestPolicyFactories`, but said factories are fixed:

```
  private readonly _requestPolicyFactories: RequestPolicyFactory[];
```

:(

In short, once I (Or Angulars DI mechanism) instantiates my service, there is no way to modify the credentials used in that instance.

I'm a big fan of DI - keeps things clean and easy to test, but for now, it looks like the only way around this problem is to inject a factory instead of an instance of the service into my components ... which I guess will work, but is just that little bit more complex than I feel this should be...

In an ideal world, the autorest MyServiceContext class generated should be able to tell the underlying `msRest.ServiceClient` that the credentials have been updated, and it should then go away and rebuild it's `_requestPolicyFactories` ... IMHO ...

... just my 0.5p ... let me know if this is not in the right place ... would appreciate some insights ...

  Jamie.

**Comment by fearthecowboy** [comment]
I don't think there is a way with the current code. 

We're crafting new generators for both typescript and c# based on the new azure core runtime, and not doing any more work on the v2 generators.
